src/Sarif.Sarifer/SarifViewerBackgroundAnalysisSink.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading;
+
+using Microsoft.Sarif.Viewer.Interop;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Utilities;
+
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.CodeAnalysis.Sarif.Sarifer
+{
+    /// <summary>
+    /// A sink for background analysis results that sends the results to the SARIF viewer through
+    /// its interop API.
+    /// </summary>
+    [Export(typeof(IBackgroundAnalysisSink))]
+    internal class SarifViewerBackgroundAnalysisSink : IBackgroundAnalysisSink
+    {
+        private readonly ReaderWriterLockSlimWrapper interopLock = new ReaderWriterLockSlimWrapper(new ReaderWriterLockSlim());
+        private SarifViewerInterop sarifViewerInterop;
+
+        /// <inheritdoc/>
+        public async Task ReceiveAsync(Stream logStream)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            SarifViewerInterop sarifViewerInterop = await GetInteropObjectAsync().ConfigureAwait(continueOnCapturedContext: true);
+
+            await sarifViewerInterop.OpenSarifLogAsync(logStream).ConfigureAwait(continueOnCapturedContext: false);
```

Comment:
>logStream [](start = 55, length = 9)

Want to discuss this with you. Mentally I have two pictures in my mind. The first is a completed "stream" versus "streaming results"

Reply:
Indeed. We've had this discussion offline and we'll migrate towards the model we discussed, using a tagger to populate the error list, and streaming out individual results (each as a self-contained SARIF file) rather than waiting to analyze the whole buffer.

---
In reply to: [519930411](https://github.com/microsoft/sarif-visualstudio-extension/pull/260#discussion_r519930411) [](ancestors = 519930411)